### PR TITLE
Add CssClass property to ValidationSummary

### LINKED ISSUE #43860: Pass css class in `<ValidationSummary />` to the generated `<ul>`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I am trying to pass additional css classes to the generated `ul` element. I am applying general formatting via the `validation-errors` css class, but I would like to be able to pass utility classes for spacing (i.e. `pt-2` to add 8px of padding to the top of the validation summary).

### Describe the solution you'd like

Update `<ValidationSummary />` to pass unmatched attributes (via splatting) to the `validation-errors` element. In the case of `class`, it should append the passed in classes to the existing `validation-errors` class.

### Additional context

For example:
```razor
<ValidationSummary class="pt-2" data-test="validation-errors" />
```

should result in:
```html
<ul class="validation-errors pt-2"  data-test="validation-errors">
    <!-- Validation Errors -->
</ul>
```

### LINKED ISSUE #46461: Update `<ValidationSummary />` to add `CssClass` and pass additional attributes to the generated DOM element
## Background and Motivation
Currently, when using `<ValidationSummary />` it is not possible to pass additional css classes and have them appended to the generated `ul` element. While most of the styling for this component is done directly on the `validation-summary` css class, there are times when it is useful to simply use a utility class (i.e. `pt-2` to apply `8px` of top padding) to the element directly. The only way to do this right now is to wrap the component in another html element and apply the styles to the wrapper element.

For example:
```html
<div class="pt-2">
  <ValidationSummary />
</div>
```

Additionally, a common pattern when creating UI automation tests is to use the `data-test` attribute to make it easier for the test to locate the element. The current version of `<ValidationSummary />` does not apply additional properties to the generated DOM element.

## Proposed API
To simplify these scenarios, `<ValidationSummary />` should pass unmatched attributes (via splatting) to the generated root `ul` element and a new property `CssClass` attribute should be added which appends the passed in css classes to the generated DOM element. This is the same pattern used in other built-in form components.

```diff
namespace Microsoft.AspNetCore.Components.Forms;

public class ValidationSummary
{
+    /// <summary>
+    /// Gets or sets the computed CSS class to be appended to the <c>ul</c> element.
+    /// </summary>
+    protected string? CssClass { get; set; }
}
```

See this open pull request: https://github.com/dotnet/aspnetcore/pull/43928

## Usage Examples
The following razor code:
```html
<ValidationSummary CssClass="pt-2" data-test="xyz-errors" />
```

Currently results in the following generated html:
```html
<ul class="validation-errors">
  <!-- Validation Errors -->
</ul>
```

After my proposal, it should result in the following generated html:
```html
<ul class="validation-errors pt-2" data-test="xyz-errors">
  <!-- Validation Errors -->
</ul>
```

## Alternative Designs
See https://github.com/dotnet/aspnetcore/issues/43860 for the full discussion.

## Risks
This is a breaking change because `ValidationSummary` is not sealed so any class that inherits it could potentially have added their own property named `CssClass`. 

This breaking change is aligning this component with other built-in components.

### PR DESCRIPTION
**Pull request by ChristopherHaws**
# Add CssClass property to ValidationSummary
API Proposal: #46461

## Description
Add `CssClass` property to `ValidationSummary` and append the contents to the resulting `class` attribute.

Fixes #43860, Closes: #46461

### CODE REVIEW

File: src/Components/Web/test/Forms/ValidationSummaryTest.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        // Act
+        var renderer = new TestRenderer();
+        var rootComponentId = renderer.AssignRootComponentId(rootComponent);
+        await renderer.RenderRootComponentAsync(rootComponentId);
+        var batch = renderer.Batches.Single();
+        var ulFrame = batch.ReferenceFrames.Single(x => x.ElementName == "ul");
+        var ulFrameIndex = Array.IndexOf(batch.ReferenceFrames, ulFrame);
+        var ulFrameSubtree = batch.ReferenceFrames.Skip(ulFrameIndex).Take(ulFrame.ElementSubtreeLength);
+
+        var ulClassFrame = ulFrameSubtree.FirstOrDefault(x =>
```
Comment (ChristopherHaws) [inline src/Components/Web/test/Forms/ValidationSummaryTest.cs:41]: I'm not used to writing tests for components without bUnit. Is this the easiest way to find the resulting `class` attribute for the generated `ul` element?

File: src/Components/Web/src/Forms/ValidationSummary.cs [hunk 2]
```diff
@@ -73,7 +78,7 @@ protected override void BuildRenderTree(RenderTreeBuilder builder)
                 first = false;
 
                 builder.OpenElement(0, "ul");
-                builder.AddAttribute(1, "class", "validation-errors");
+                builder.AddAttribute(1, "class", string.IsNullOrWhiteSpace(CssClass) ? "validation-errors" : $"validation-errors {CssClass}");
```
Comment (ChristopherHaws) [inline src/Components/Web/src/Forms/ValidationSummary.cs:81]: As I state [here](https://github.com/dotnet/aspnetcore/issues/43860#issuecomment-1244145778), other components seem to have `CssClass` as a get only property and use `AttributeUtilities.CombineClassNames` in `OnParametersSet` to append the class name passed in via `AdditionalAttributes`. Let me know if I should update this code to do that as well.

Thanks! :)

Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/ValidationSummary.cs:81]: Yes, sorry, you're right. It would be better to align more with the pattern we have elsewhere of extracting the `class` value from `AdditionalAttributes`, with `CssClass` being a readonly property that combines the generated class with any supplied by the developer. Sorry for the misleading advice previously.

File: src/Components/Web/test/Forms/ValidationSummaryTest.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Web/test/Forms/ValidationSummaryTest.cs:41]: Typically we don't write unit tests for this kind of component rendering. Instead we write end-to-end (E2E) tests that rendering into a real browser and are automated with Selenium.

If you have a look at `src\Components\test\E2ETest\Tests\FormsTest.cs` and `TypicalValidationComponent.razor` you'll see some examples. It should be reasonably straightforward to add a further case following this pattern to show that `ValidationSummary` renders a supplied class name. These E2E tests can be executed as normal tests via `dotnet test` or the VS test window, but if you run the whole suite of them it can take a while, like 15 minutes or more.

Comment (ChristopherHaws) [reply src/Components/Web/test/Forms/ValidationSummaryTest.cs]: Sounds good. I switched this to an E2E test. Had a hell of a time getting selenium to run because the web driver requires v103 of chrome but the latest version of chrome is v105 and there is no way to download an old version of chrome. Ended up having to find a build of chromium for v103. You may want to add this info to the build docs. I remember now why I switched away from selenium and started using playwright instead :)

**Comment by ChristopherHaws** [comment]
@SteveSandersonMS I updated the test to be an E2E test. I had a hell of a time getting selenium to work because I have v105 of chrome installed but the web driver used by the E2E tests is still v103 and chrome doesn't supply old versions of chrome. I had to lookup the build number of chromium that corresponds to v103 of chrome and use that instead. I remember now why I stopped using selenium and switched to playwright. :)

I think this PR is ready for review now. Thanks!

**Comment by mkArtakMSFT** [comment]
Thanks for your PR, @ChristopherHaws.
@SteveSandersonMS could you please review this? Thanks!

**Comment by mkArtakMSFT** [comment]
@Nick-Stanton reassigning this to you as @SteveSandersonMS has been busy with other things.

File: src/Components/Web/src/PublicAPI.Shipped.txt [hunk 3]
```diff
@@ -140,6 +140,8 @@ Microsoft.AspNetCore.Components.Forms.ValidationSummary.AdditionalAttributes.get
 Microsoft.AspNetCore.Components.Forms.ValidationSummary.AdditionalAttributes.set -> void
 Microsoft.AspNetCore.Components.Forms.ValidationSummary.Model.get -> object?
 Microsoft.AspNetCore.Components.Forms.ValidationSummary.Model.set -> void
+Microsoft.AspNetCore.Components.Forms.ValidationSummary.CssClass.get -> string?
+Microsoft.AspNetCore.Components.Forms.ValidationSummary.CssClass.set -> void
```
Comment (Nick-Stanton) [inline src/Components/Web/src/PublicAPI.Shipped.txt:144]: Please move these to the PublicAPI.Unshipped.txt file in the same directory.

**Review by Nick-Stanton — CHANGES_REQUESTED** [review]
Hi @ChristopherHaws, thank you for your contribution and apologies for the delay. I made one small request, but otherwise the changes look good.

Since your PR contains new API, you will need to create an API Proposal to go along with your issue. This is done by opening a new issue using the API Proposal template. More information about the process is [here](https://github.com/dotnet/aspnetcore/blob/main/docs/APIReviewProcess.md). Once the proposal is ready, I will "champion" your changes on your behalf during review. Please let me know if you have any questions.

File: src/Components/Web/src/PublicAPI.Shipped.txt (hunk 3 above)
Comment (ChristopherHaws) [reply src/Components/Web/src/PublicAPI.Shipped.txt:144]: Sounds good, I will work on this after work today! :) Thanks!

Comment (Nick-Stanton) [reply src/Components/Web/src/PublicAPI.Shipped.txt:144]: > Should I update my existing issue with the proposal template or should I make a new issue and link both?

Make a new issue and link anything relevant please 😃

Comment (ChristopherHaws) [reply src/Components/Web/src/PublicAPI.Shipped.txt]: Hi @Nick-Stanton. Sorry it took me a bit to get to this. I have been super busy at work. :)

I moved the new api's to the `PublicAPI.Unshipped.txt` file and created an api proposal for this change here: https://github.com/dotnet/aspnetcore/issues/46461

**Comment by Nick-Stanton** [comment]
@ChristopherHaws thanks for taking the time to draft your proposal. It looks like the commit that moved your new API caused some new build failures. I imagine CI will be clean once those are fixed.

While investigating these changes deeper to prepare for API review, I had a couple questions:

1. First, I was looking at your example in #43860. If you were to pass `class="validation-errors pt-2"`, wouldn't you get your desired output? I ask because if that scenario works and this feature hasn't been asked for elsewhere, it may be hard to argue for this change. Example:
```html
<ValidationSummary class="validation-errors pt-2" data-test="validation-errors" />
```
2. This implementation has `ValidationSummary` modifying its own CssClass parameter, which you were right to ask about here https://github.com/dotnet/aspnetcore/issues/43860#issuecomment-1244145778. We [usually guide against doing this](https://learn.microsoft.com/aspnet/core/blazor/components/?view=aspnetcore-7.0#overwritten-parameters). Is there a way to achieve this functionality without directly changing `CssClass`?

**Comment by ChristopherHaws** [comment]
@Nick-Stanton Hmm, not sure why the build would be failing. Unfortunately I don't have time to look at that this week, I am super busy at work and have a theater show I am running at night.

1. The existing `<ValidationSummary />` component adds `validation-errors`, not the end user.
2. If I remember correctly (it's been almost 6 months since I wrote this code), the CssClass property needs to contain the class names being appended to the component which includes the built-in `validation-errors` css class. 

The only built-in component I found which allows users to set `CssClass` is [NavLink](https://github.com/dotnet/dotnet/blob/febdb074f5ada900f6ef37e35f297b0fced54e8c/src/aspnetcore/src/Components/Web/src/Routing/NavLink.cs#L81)) which also appends it's built-in css class to the property.

Most of the built-in components don't allow you to pass CssClass ([it is get only](https://github.com/dotnet/dotnet/blob/b8bc661a7429baa89070f3bee636b7fbc1309489/src/aspnetcore/src/Components/Web/src/Forms/InputBase.cs#L164)).

I don't know the exact reason that the built-in css class names need to be in this property but I am guessing it is one or many of the following:
- Performance. If you append the string only one time when the parameter is set, you avoid string allocations every time the component gets rendered/re-rendered
- If someone has a ref to the component, they could then get the `CssClass` and it would actually contain all of the css classes, not just the ones the user added
- When writing tests, this would allow you to assert that the css classes are correct without needing to validate it using the DOM

**Comment by SteveSandersonMS** [comment]
> The only built-in component I found which allows users to set CssClass is [NavLink](https://github.com/dotnet/dotnet/blob/febdb074f5ada900f6ef37e35f297b0fced54e8c/src/aspnetcore/src/Components/Web/src/Routing/NavLink.cs#L81)) which also appends it's built-in css class to the property.

Ah, checking this more closely, I see that the `CssClass` property you're adding is not actually a `[Parameter]` property so is fine to mutate. (Apologies to @Nick-Stanton - I gave an incorrect suggestion about this.). However I'm not sure it should be settable by subclasses, since their changes would be overwritten every time `OnParametersSet` runs which would be very confusing.

> The existing <ValidationSummary /> component adds validation-errors, not the end user.

That's right, `validation-errors` is the *default* CSS class attribute value. However it's intentional that developers can overwrite this with any other that they want. For example, maybe their styling system doesn't use that CSS class name, so they instead want to use `<ValidationErrors class="my-errors">`. It's intended as a feature that you can replace the class name. It's the same with `ValidationMessage` too.

The change in the PR, if I'm reading correctly, would prevent that and force people to have `validation-errors` whether or not they want it. I know in most cases that won't cause them any problem, but I'm not really sold on the idea that a lot of people are going to be using `class` to represent *additional* class names as opposed to thinking they are setting the entire `class` attribute. Especially given how this would then be inconsistent with `ValidationMessage`.

The reason why we combine classes dynamically in `NavLink` is that `NavLink` adds and removes the `active` class based on navigation state. There has to be *some* dynamic adding/removing of this class since that's the whole point of `NavLink`. However we still don't force any specific hardcoded class name since you can set `ActiveClass` to an arbitrary value.

**In summary** it looks as if we already have the desired functionality, i.e., the ability to customize the CSS class name and add arbitrary extra ones just by specifying `class`. I don't think we'd want to force people to have `validation-errors`. Hope that seems reasonable. If I'm missing something important, please let me know!

**Comment by ChristopherHaws** [comment]
@SteveSandersonMS The current behavior of `ValidationSummary` is that you simply cannot override the `class` attribute at all. If you pass a `class`, it is simply ignored. Here are two examples I just tested:
```html
<!-- <ValidationSummary /> -->
<ul class="validation-errors">
    <li class="validation-message">...</li>
</ul>

<!-- <ValidationSummary class="test" data-test="test" /> -->
<ul data-test="test" class="validation-errors">
    <li class="validation-message">...</li>
</ul>
```

I see your point about it being confusing as to whether it replaces or appends the class though. I am coming to blazor after using React for many years where it is pretty common for the classes you pass in to be appended, so I would have assumed that it would be appended. [Here are the docs for FluentUI React](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield#implementation) where you can see that `className` is an `Optional class name that is added to the container of the component.`

I get that this is not React, so I looked at one of the top Blazor Component libraries to see what they do. [MudBlazor](https://www.mudblazor.com/api/field#properties) defines `Class` as `User class names, separated by space.` - The key word here being that this is the "User"'s class names.

**Comment by SteveSandersonMS** [comment]
@ChristopherHaws Can you provide a repro using an online REPL?

I just tried it and it works for me: https://blazorrepl.telerik.com/QRumvqaN29a37x4K49

image

**Comment by ChristopherHaws** [comment]
Interesting, I had used [BlazorFiddle](https://blazorfiddle.com/s/6sqte5xq) to try it. I guess they are on an old version of Blazor. It's been a while since I was dealing with this issue in our project (we ended up just wrapping `ValidationSummary` in our own component).
image

**Comment by Nick-Stanton** [comment]
@ChristopherHaws I just looked into it and it looks like they're using a _really_ old version of Blazor.
image
Apologies for not discovering the functionality earlier in review. Do you still have interest in pursuing this PR? It seems like the ask is already met.

**Comment by mkArtakMSFT** [comment]
Closing as this is already supported.
